src/ffishjs.cpp

```diff
@@ -263,10 +263,59 @@ class Board {
     return pos.game_ply();
   }
 
+  std::string result() const {
+    return result(false);
+  }
+
+  std::string result(bool claim_draw) const {
+    Value result;
+    bool gameEnd = false;
+    if (is_insufficient_material()) {
+      gameEnd = true;
+      if (pos.material_counting())
+        result = pos.material_counting_result();
+      else
+        result = VALUE_DRAW;
```

Comment:
I think this whole function should be reviewed to make sure I am complying with all of the rules that affect variants, since I don't know most of them :) But this part in particular - does this make sense? I'm not exactly sure when material counting comes into play. I figure that most of the weird end-game conditions are handled by `pos.is_immediate_game_end()` but I want to make sure that this is correct.

I will probably need more help when I write the `board.outcome()` function, because there, I actually need to know why the game ended.

@ianfab

Reply:
I would like to avoid duplicating variant logic as much as possible, since it is very complex, and it should mostly be defined in `Position` methods as the core game logic. Therefore I think the draw result adjudication should be moved to a `Position` method, which anyway would simplify the code, since this very same logic regarding material counting already is duplicated in three places (my bad^^). It could e.g. be something like `Value Position::draw_result(int ply = 0)`.

The outcome method is somewhat problematic, since implementing it would either duplicate logic, or complicate the position adjudication if integrated there. The game end methods could of course return an enum instead of a bool, which likely should not have any impact on performance or functionality, although I am still not sure if this added complexity is acceptable considering that the core engine does not need this information at all. So I am a bit undecided on this topic at the moment. Also @gbtami might be interested in this discussion since pychess-variants currently has custom logic outside of pyffish to determine the type of game end for displaying purposes.

Reply:
Yea, pychess has its own code (with lots of variant specific parts) to determine game results. See https://github.com/gbtami/pychess-variants/blob/master/server/game.py#L516

Reply:
These are great comments, thanks guys.

Regarding draw adjudication, I figured out that my current method doesn't even work correctly (if the variant is set to armageddon and you get a stalemate, the result is `1/2-1/2` instead of `0-1`). That might be easily fixable in this case, just set the result to `VALUE_DRAW` on insufficient material, then before returning, check if the current value is `VALUE_DRAW` and if so, factor in material counting. Of course, as you said, it may be nicer to have a function somewhere in `Position`, I'd be happy to take a crack at it though I'm not sure I'll do it right :)

I checked out the pychess-variants code, and it's definitely more complicated than I thought it would be :smile: I'm also realizing it's not fully complete, for instance it just checks if the variant name is "atomic" when seeing if a pseudo-royally checkmated piece should count as checkmate rather than stalemate. How can we know that generically, for all similar variants? Of course, that's fine for pychess.org since that's the only variant that handles that case. But if we want it in a generic library, there is extra complexity.

Also, there are probably some refactors we would want to make for optional draws. AFAIK, n-fold repetition and n-move rule (and others) are all implemented in one lumped-together method `Position::is_optional_game_end()`. How can we check which of those end conditions was reached unless we re-implement them ourselves? My guess is it would be best to split them into different functions (maybe `Position::is_optional_game_end_n_fold()`, `Position::is_optional_game_end_n_moves()`, etc), then check for all of those functions in `Position::is_optional_game_end()`. That way, that function still works as intended, and we can now check for the individual conditions when we need them (like in ffish.js). That's an idea.

Reply:
> The outcome method is somewhat problematic, since implementing it would either duplicate logic, or complicate the position adjudication if integrated there. The game end methods could of course return an enum instead of a bool, which likely should not have any impact on performance or functionality, although I am still not sure if this added complexity is acceptable considering that the core engine does not need this information at all. So I am a bit undecided on this topic at the moment. Also @gbtami might be interested in this discussion since pychess-variants currently has custom logic outside of pyffish to determine the type of game end for displaying purposes.

I share your concern about duplicate code functionality.
Do you see a way of moving some of these methods to e.g. [apiutil.h](https://github.com/ianfab/Fairy-Stockfish/blob/9205b0839984c5c9f5afa5ce6022de68b9a3e5a7/src/apiutil.h) or to let them depend on already existing methods in a more principled way? I think these kinds of new methods would greatly improve the user experience of the library.

Reply:
I had the idea to add `Termination &outcome` to methods like `Position::is_immediate_game_end`, right next to the `Value &result`. That way, in the function where it's being asked, we can optionally output the `Termination` at the same time as the `Value`. So, something like:

```c++
bool Position::is_immediate_game_end(Value& result, int ply) const {
    Termination outcome;
    return Position::is_immediate_game_end(result, outcome, ply);
}

bool Position::is_immediate_game_end(Value& result, Termination& outcome, int ply) const {
    // ...
    outcome = VARIANT_END; // or CHECKMATE, etc
    // ...
}
```

Apply a similar thing to all functions that output a `Value` representing the end of the game.

Does that sound feasible/reasonable?